File: python/Inserción.py
import time

def main():
    ruta_archivo = "C:\\Users\\Giuliana Espinoza\\Documents\\Análisis de algoritmos\\DatosCargados.txt"
    
    lista_numeros = leer_datos_desde_archivo(ruta_archivo)
    if lista_numeros is None:
        print("Error al leer el archivo.")
        return

    limite_datos = min(5000, len(lista_numeros))
    datos_a_ordenar = lista_numeros[:limite_datos]  # Tomar solo una parte de los datos

    inicio_tiempo = time.time()
    lista_ordenada = ordenar_por_insercion(datos_a_ordenar)
    fin_tiempo = time.time()

    tiempo_segundos = fin_tiempo - inicio_tiempo
    print(f"Tiempo de ejecución para {limite_datos} datos: {tiempo_segundos:.6f} segundos")

    if limite_datos <= 50:
        print("Lista ordenada:", lista_ordenada)

def leer_datos_desde_archivo(nombre_archivo):
    lista_temporal = []
    try:
        with open(nombre_archivo, 'r') as archivo:
            for linea in archivo:
                lista_temporal.append(int(linea.strip()))  # Leer cada número y agregarlo a la lista
    except FileNotFoundError:
        print(f"Error: No se encontró el archivo {nombre_archivo}")
        return None
    return lista_temporal

def ordenar_por_insercion(lista_numeros):
    for i in range(1, len(lista_numeros)):  # Comienza desde el segundo elemento
        valor_actual = lista_numeros[i]
        indice_comparacion = i - 1

        # Mover los elementos que son mayores que el valor_actual una posición a la derecha
        while indice_comparacion >= 0 and lista_numeros[indice_comparacion] > valor_actual:
            lista_numeros[indice_comparacion + 1] = lista_numeros[indice_comparacion]
            indice_comparacion -= 1

        # Colocar el valor actual en su posición correcta
        lista_numeros[indice_comparacion + 1] = valor_actual

    return lista_numeros

if __name__ == "__main__":
    main()

File: Java/src/ordenamientoinsercion/Inserción.java
package Java.src.ordenamientoinsercion;

import java.io.*;
import java.util.*;

public class Inserción {

    public static void main(String[] args) {
        // Ruta correcta para Visual Studio Code, con ruta relativa
        String rutaArchivo = "DatosCargados.txt";  // Ajusta la ruta según la ubicación del archivo
        
        int[] listaNumeros = leerDatosDesdeArchivo(rutaArchivo);
        if (listaNumeros == null) {
            System.out.println("Error al leer el archivo.");
            return;
        }

        int limiteDatos = Math.min(10000, listaNumeros.length);
        int[] datosAOrdenar = Arrays.copyOfRange(listaNumeros, 0, limiteDatos); // Tomar solo una parte de los datos

        long inicioTiempo = System.nanoTime();
        int[] listaOrdenada = ordenarPorInsercion(datosAOrdenar);
        long finTiempo = System.nanoTime();

        double tiempoSegundos = (finTiempo - inicioTiempo) / 1e9;
        System.out.println("Tiempo de ejecución para " + limiteDatos + " datos: " + tiempoSegundos + " segundos");

        if (limiteDatos <= 50) {
            System.out.println("Lista ordenada: " + Arrays.toString(listaOrdenada));
        }
    }

    public static int[] leerDatosDesdeArchivo(String nombreArchivo) {
        List<Integer> listaTemporal = new ArrayList<>();
        try (Scanner scanner = new Scanner(new File(nombreArchivo))) {
            while (scanner.hasNextInt()) {
                listaTemporal.add(scanner.nextInt());
            }
        } catch (FileNotFoundException e) {
            System.out.println("Error: No se encontró el archivo " + nombreArchivo);
            return null;
        }

        int[] listaNumeros = new int[listaTemporal.size()];
        for (int i = 0; i < listaTemporal.size(); i++) {
            listaNumeros[i] = listaTemporal.get(i);
        }
        return listaNumeros;
    }

    public static int[] ordenarPorInsercion(int[] listaNumeros) {
        for (int i = 1; i < listaNumeros.length; i++) {
            int valorActual = listaNumeros[i];
            int indiceComparacion = i - 1;

            // Mover los elementos mayores que el valorActual una posición a la derecha
            while (indiceComparacion >= 0 && listaNumeros[indiceComparacion] > valorActual) {
                listaNumeros[indiceComparacion + 1] = listaNumeros[indiceComparacion];
                indiceComparacion--;
            }

            // Colocar el valorActual en la posición correcta
            listaNumeros[indiceComparacion + 1] = valorActual;
        }
        return listaNumeros;
    }
}
